modules/apps/static/portal-osgi-web/portal-osgi-web-portlet-container-test/src/testIntegration/java/com/liferay/portal/osgi/web/portlet/container/test/BasePortletContainerTestCase.java

```diff
@@ -87,6 +90,33 @@ protected void registerService(
 				properties);
 
 		serviceRegistrations.add(serviceRegistration);
+
+		String portletName = (String)properties.get("javax.portlet.name");
+
+		try {
+			Filter filter = FrameworkUtil.createFilter(
+				String.format(
+					"(&(objectClass=%s)(javax.portlet.name=%s))",
+					com.liferay.portal.kernel.model.Portlet.class.getName(),
+					portletName));
+
+			ServiceTracker<?, ?> trackInstalledPortlet = new ServiceTracker<>(
+				bundleContext, filter, null);
+
+			trackInstalledPortlet.open();
+
+			trackInstalledPortlet.waitForService(5000);
```

Comment:
@rotty3000 I would need more context info for this.
A blind time waiting can never be good. On CI timing is very unpredictable. Even this 5000ms works most times, eventually we could see a failure again.
Can you explain a bit about what this race condition is? And better if you have a way to force it to happen. I can try to find a more reliable way to fix it.

Reply:
I'll just remove the commit since I didn't make the portlet registration async in the end.

Reply:
Great, that makes a lot sense! Can you resend the pull without it, and run through CI again to be sure? Thanks @rotty3000 !